[shopfloor] rework datamodel

guewen opened the pull request:
This is a simplification of the model, which was hard to understand
and configure.
We had too many options to filter the visibility of menus / processes,
which made things hard to debug. Some models were overlapping in term
of concept and fonctionality.

* Processes and menus are redundant, remove processes and keep only menus
* Remove shopfloor groups, we can restrict the visibility of a menu
  using profiles. Later, only a manager should be able to change the
  profile on the devices.
  Additionally, we would still be able to use ir.rules if needed.
* On the menus, we'll be able to select several picking types. When a scenario has to know a single
  picking type, for instance to create a new move, we'll use a dedicated
  many2one for this purpose.

Details in commits

Commit:
Rework shopfloor datamodel

This is a simplification of the model, which was hard to understand
and configure.
We had too many options to filter the visibility of menus / processes,
which made things hard to debug. Some models were overlapping in term
of concept and fonctionality.

* Processes and menus are redundant, remove processes and keep only menus
* Remove shopfloor groups, we can restrict the visibility of a menu
  using profiles. Later, only a manager should be able to change the
  profile on the devices.
  Additionally, we would still be able to use ir.rules if needed.
* On the menus, we'll be able (still to implement in the services)
  to select several picking types. When a scenario has to know a single
  picking type, for instance to create a new move, we'll use a dedicated
  many2one for this purpose.

Commit:
checkout: modify lookup by package

The test CheckoutScanCase.test_scan_document_error_different_picking_type_package
was no longer passing because was returning a "barcode not
found" instead of "you cannot move this using this menu".

The picking type should not be checked at this point, as it is checked
in _select_picking() (called at the end of the method), which will
return the appropriate message.

Note: replaced the search by a field to reuse the cache if we had such
data in cache.

Commit:
backend: support multiple picking types in services

Except for the cases where we can create a move, a single menu should
be able to work on several picking types.

Later, some menus will have a flag "create_move", which add an extra
step in the scenario. In this case, we'll still have a selection of
several picking types on the menu, but an additional M2o to select the
picking type used for creation of moves.

Commit:
run black

guewen commented:
@simahawk In theory, it shouldn't have any impact on the front-end side (except the changes I already did). I tested the config / profiles / menus of the app, which still work, but didn't run any scenario yet.

simahawk commented:
I'm about to review. Can you review this 1st https://github.com/camptocamp/wms/pull/9 ?
You can pay attention to backend part only as I fixed few issues coming from recent changes (tests were not updated as I did it in a rush).
Would be nice to merge and rebase here then.

simahawk commented on shopfloor/models/stock_quant_package.py:
```diff
@@ -4,6 +4,13 @@
 class StockQuantPackage(models.Model):
     _inherit = "stock.quant.package"
 
+    move_line_ids = fields.One2many(
```
is this needed if we rely on https://github.com/camptocamp/wms/pull/9/commits/0ac9e9f4db76b3df00ef42d8563f4430984abfb1#diff-8669ee167ce03adadc1c12257e201f58R41 ?

guewen replied:
I try to limit the number of calls to search which never use the cache

simahawk commented on shopfloor/__manifest__.py:
```diff
@@ -25,9 +25,7 @@
     ],
     "data": [
         "security/ir.model.access.csv",
-        "views/shopfloor_operation_group.xml",
```
great! Is the update of our integration module on your radar?

guewen replied:
I'll do it

simahawk commented on shopfloor/services/checkout.py:
```diff
@@ -85,17 +85,19 @@ def scan_document(self, barcode):
-                            ("package_id", "=", package.id),
-                            ("picking_id.picking_type_id", "=", self.picking_type.id),
-                        ]
+                pickings = package.move_line_ids.filtered(
```
TBH I prefer the search, I factored it out to a specific helper method in my PR

simahawk commented on shopfloor/services/menu.py:
```diff
@@ -21,21 +20,18 @@ class ShopfloorMenu(Component):
+                [
+                    "|",
+                    ("profile_ids", "=", False),
+                    ("profile_ids", "in", self.work.profile.id),
```
```suggestion
                    ("profile_ids", "in", self.work.profile.ids),
```

simahawk commented on shopfloor/services/service.py:
```diff
@@ -24,19 +24,21 @@ class BaseShopfloorService(AbstractComponent):
+        """Return picking types for the menu and profile"""
+        # TODO make this a lazy property or computed field avoid running the
+        # filter every time?
+        picking_types = self.work.menu.picking_type_ids.filtered(
```
shall we cache this by menu/profile?

guewen replied:
yes, would be worth it I think, but for now I keep it as TODO

simahawk commented on shopfloor/services/single_pack_putaway.py:
```diff
@@ -85,8 +85,12 @@ def _response_for_start_success(self, move_line, pack):
         """Scan a pack barcode"""
-
-        picking_type = self.picking_type
+        # TODO we have to rework this and single_pack_transfer, 'pack putaway'
```
:dizzy_face:

simahawk commented on shopfloor/tests/test_checkout_cancel_line.py:
```diff
@@ -85,7 +85,10 @@ def test_cancel_package_ok(self):
-            data={"picking": self._stock_picking_data(picking, done=True)},
+            data={
+                "picking": self._stock_picking_data(picking, done=True),
+                "all_processed": False,
```
doh, sorry, I fixed these already :sweat_smile:

Commit:
backend: hide menus with picking types of different warehouse

guewen commented on shopfloor/services/menu.py:
```diff
@@ -21,31 +20,35 @@ class ShopfloorMenu(Component):
-            # This data model is a bit messy :/
-            op_group_domain = [
-                ("operation_group_ids", "in", profile.operation_group_ids.ids),
+        # FIXME exclude menus if they have no picking types
+        # in warehouse matching the profile
```
```suggestion
```

guewen commented:
@simahawk rebased

Commit:
Fix build

The version of stock_move_common_dest installed by pip misses
a correction done in https://github.com/OCA/stock-logistics-warehouse/pull/808

Commit:
Add MQT_DEP in the OCA/OCB build

It works on the odoo build
